feat: remove the limit of term splitting

hualet opened the pull request:
a feature request from here: https://bbs.deepin.org/post/277237

hualet committed:
feat: remove the limit of term splitting

a feature request from here: https://bbs.deepin.org/post/277237

hualet commented:
I have read the CLA Document and I hereby sign the CLA.

rb-union commented:
/check clacheck/CLAssistant

rb-union commented:
https://github.com/linuxdeepin/deepin-terminal/blob/7a69abbf9cfbaf5be52531aae967d95a72fe3225/src/views/termwidget.cpp#L126
这里有设置最小的大小限制，小窗口下显示分屏多的话，下面的分屏终端显示会被遮盖。

BLumia commented:
recheck

hualet commented:
I have read the CLA Document and I hereby sign the CLA.

hualet committed:
fix: no term gets focus after closing one

fix no term gets focus after closing one, which will cause crashes
if you hit Alt+Q in sequence.

hualet committed:
fix: change minimum size set on TermWidget

original minimum size relies on the limit of one can only split
the terminal once, since we removed the limits, the minimum size
should change accordingly

hualet commented:
- [x] 在中间某个term上分屏，新终端出现的位置不对
- [x] 拖动分割线以后，终端分屏的尺寸不对
- [x] 在无法分屏的时候（如何判断？），隐藏或者禁用分屏菜单项

hualet committed:
fix: new term not showing in the right position

fix the bug that new term not showing right after the old one after
the splitting

hualet committed:
fix: new term resized wrongly

fix the bug that new term has wrong size after splitting

hualet committed:
fix: disable splitting menu if there's no room

disable splitting menu when there's no enough room for splitting

hualet committed:
fix: shortcut can still do splitting on room left

fix shortcut still work even when there's no room for splitting

hualet committed:
fix: crash of unfound size

createSubSplit has side effects, it removes the term form its
parent, so find size of it in QSplitter::sizes crashes.

hualet commented:
@rb-union  @ArchieMeng  done!

ArchieMeng commented on src/views/termwidget.h:
```diff
@@ -65,6 +65,10 @@ class TermWidget : public QTermWidget
 public:
+    // 9:6
+    static const int MIN_WIDTH = 180;
+    static const int MIN_HEIGHT = 120;
```
It can be private if CanSplit is a friend function or an object function.

ArchieMeng commented on src/views/termwidget.h:
```diff
@@ -467,4 +464,35 @@ private slots:
     bool m_remotePasswordIsReady = false;
 };
 
+static bool CanSplit(TermWidget *term, Qt::Orientation ori) {
```
It would be better if this is a public member function of TermWidget. Is a static function required here? If so, then why is it required?

ArchieMeng reviewed (changes requested):
Basically LGTM.
Additionally, a commit message is required by https://github.com/linuxdeepin/deepin-terminal/actions/runs/10635689078/job/29485856049?pr=353

hualet committed:
chore: some code refactory

- change CanSplit to a public funtion of TermWidget
- change MIN_WIDTH and MIN_HEIGHT of TermWidget to private

ArchieMeng commented on src/main/mainwindow.cpp:
```diff
@@ -1131,7 +1131,7 @@ inline void MainWindow::slotShortcutHorizonzalSplit()
     if (Service::instance()->isCountEnable()) {
         TermWidgetPage *page = currentPage();
-        if (page && CanSplit(page->currentTerminal(), Qt::Vertical)) {
+        if (page && TermWidget::canSplit(page->currentTerminal(), Qt::Vertical)) {
```
I meant "convert canSplit to object function". So that, the way of calling it becomes
```C++
page->currentTerminal()->canSplit(Qt::Vertical)
```

ArchieMeng commented on src/views/termwidget.cpp:
```diff
@@ -1011,6 +1011,38 @@ bool TermWidget::isInRemoteServer()
     return false;
 }
 
+bool TermWidget::canSplit(TermWidget *term, Qt::Orientation ori) {
```
I meant "convert canSplit to object function".

ArchieMeng reviewed (approved):
+1
